tests/test_controlsocket.py

```diff
@@ -0,0 +1,155 @@
+            )
+        )
+        control_socket.add_metrics_user('juju-metrics-r0', 'passwd')
+
+    def test_add_metrics_user_fail(self):
+        mock_opener = mockOpener(self)
+        control_socket = Client('fake_socket_path', opener=mock_opener)
+
+        mock_opener.expect(
+            url='http://localhost/metrics-users',
+            method='POST',
+            body=r'{"username": "juju-metrics-r0", "password": "passwd"}',
+
+            error=urllib.error.HTTPError(
+                url='http://localhost/metrics-users',
+                code=409,
+                msg='',
+                hdrs=None,
+                fp=io.BytesIO(br'{"error":"user \"juju-metrics-r0\" already exists"}'),
+            )
+        )
+
+        self.assertRaisesRegex(
+            APIError,
+            r'user "juju-metrics-r0" already exists',
+            control_socket.add_metrics_user, 'juju-metrics-r0', 'passwd',
+        )
+
+    def test_remove_metrics_user_success(self):
+        mock_opener = mockOpener(self)
+        control_socket = Client('fake_socket_path', opener=mock_opener)
+
+        mock_opener.expect(
+            url='http://localhost/metrics-users/juju-metrics-r0',
+            method='DELETE',
+            body=None,
+
+            response=mockResponse(
+                headers=mockHeaders(content_type='application/json'),
+                body=r'{"message":"deleted user \"juju-metrics-r0\""}'
+            )
+        )
+        control_socket.remove_metrics_user('juju-metrics-r0')
+
+    def test_remove_metrics_user_fail(self):
+        mock_opener = mockOpener(self)
+        control_socket = Client('fake_socket_path', opener=mock_opener)
+
+        mock_opener.expect(
+            url='http://localhost/metrics-users/juju-metrics-r0',
+            method='DELETE',
+            body=None,
+
+            error=urllib.error.HTTPError(
+                url='http://localhost/metrics-users/juju-metrics-r0',
+                code=404,
+                msg='',
+                hdrs=None,
+                fp=io.BytesIO(br'{"error":"user \"juju-metrics-r0\" not found"}'),
+            )
+        )
+
+        self.assertRaisesRegex(
+            APIError,
+            r'user "juju-metrics-r0" not found',
+            control_socket.remove_metrics_user, 'juju-metrics-r0',
+        )
+
+    def test_connection_error(self):
+        mock_opener = mockOpener(self)
+        control_socket = Client('fake_socket_path', opener=mock_opener)
+
+        mock_opener.expect(
+            url='http://localhost/metrics-users',
+            method='POST',
+            body=r'{"username": "juju-metrics-r0", "password": "passwd"}',
+
+            error=urllib.error.URLError('could not connect to socket')
+        )
+
+        self.assertRaisesRegex(
+            ConnectionError,
+            'could not connect to socket',
+            control_socket.add_metrics_user, 'juju-metrics-r0', 'passwd',
+        )
+
+
+class mockOpener:
```

Comment:
Naming nit: Python doesn't really use `camelCase` style. For classes, it would be `MockOpener` and the like. See https://peps.python.org/pep-0008/#class-names